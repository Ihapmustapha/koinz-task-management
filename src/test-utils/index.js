export { default as MockedProvider } from "./mockedProvider";
