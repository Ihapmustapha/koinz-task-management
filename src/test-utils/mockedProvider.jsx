import React from "react";
import PropTypes from "prop-types";
import { Provider } from "react-redux";
import { configureStore } from "../store";

const MockedProvider = ({ initialState, children }) => (
  <Provider store={configureStore(initialState)}>{children}</Provider>
);

MockedProvider.propTypes = {
  initialState: PropTypes.shape({}).isRequired,
  children: PropTypes.arrayOf(PropTypes.node).isRequired,
};

export default MockedProvider;
