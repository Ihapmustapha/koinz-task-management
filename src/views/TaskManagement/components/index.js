export { default as Task } from "./Task";
export { default as TaskForm } from "./TaskForm";
export { default as Column } from "./Column";
