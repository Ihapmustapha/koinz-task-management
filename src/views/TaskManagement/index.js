export { default as TaskManagement } from "./TaskManagement";
