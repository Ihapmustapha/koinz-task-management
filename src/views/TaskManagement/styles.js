const styles = (theme) => ({
  columnPaper: {
    width: "275px",
    height: "85vh",
    padding: "20px",
    margin: "15px 20px 0px 0px",
  },
  columnTitle: {
    fontWeight: "bold",
  },
  titleDivider: {
    height: "2px",
    width: "100%",
    color: "black",
    marginTop: "10px",
  },
  tasksListContainer: {
    minHeight: "500px",
    overflow: "hidden",
    backgroundColor: "black",
  },
  addButton: {
    backgroundColor: "darkslateblue",
  },
});

export default styles;
