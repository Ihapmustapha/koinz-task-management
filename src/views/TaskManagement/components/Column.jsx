import React, { useState } from "react";
// prop-types
import PropTypes from "prop-types";
// redux
import { compose } from "redux";
import { connect } from "react-redux";
// react-beautiful-dnd
import { Droppable } from "react-beautiful-dnd";
// components
import {
  Grid,
  Paper,
  Divider,
  Tooltip,
  Fab,
  Typography,
  withStyles,
} from "@material-ui/core";
import Task from "./Task";
import TaskForm from "./TaskForm";
import { ContainerModal } from "../../../components";
// redux actions
import * as actions from "../../../store/actions";
// styles
import styles from "../styles";
// icons
import AddIcon from "../../../assets/plus.svg";

const Column = ({ column, tasks, classes, openModal }) => {
  const { title, id } = column;

  const [modalChildren, setModalChildren] = useState(null);
  const [formType, setFormType] = useState("");

  const handleTaskCardClick = (selectedTaskId) => (e) => {
    const selectedTaskDetails = tasks.filter(
      (element) => element.id === selectedTaskId
    )[0];
    if (selectedTaskDetails) {
      setFormType("update");
      setModalChildren(
        <TaskForm formType="update" selectedTaskDetails={selectedTaskDetails} />
      );
      openModal();
    }
  };

  const handleNewTaskButtonClick = () => {
    setFormType("add");
    setModalChildren(<TaskForm formType="add" selectedTaskDetails={null} />);
    openModal();
  };

  return (
    <>
      {/* <Modal /> */}
      {modalChildren && (
        <ContainerModal
          title={formType === "add" ? "Add New Task" : "Update Task"}
        >
          {modalChildren}
        </ContainerModal>
      )}
      <Grid
        component={Paper}
        container
        item
        direction="column"
        justify="center"
        alignItems="center"
        elevation={3}
        className={classes.columnPaper}
      >
        <Grid
          item
          xs={12}
          lg={12}
          container
          direction="row"
          justify="center"
          alignItems="flex-start"
        >
          <Grid item xs={12}>
            <Typography variant="body2" className={classes.columnTitle}>
              {title}
            </Typography>
            <Divider className={classes.titleDivider} />
          </Grid>
          <Grid item>
            <Droppable droppableId={id}>
              {(provided) => (
                <div
                  className={classes.tasksListContainer}
                  ref={provided.innerRef}
                  {...provided.droppableProps}
                >
                  {tasks.length > 0 &&
                    tasks.map(
                      (task, index) =>
                        task && (
                          <Task
                            key={task.id}
                            id={task.id}
                            index={index}
                            task={task}
                            onClick={handleTaskCardClick}
                          />
                        )
                    )}
                  {provided.placeholder}
                </div>
              )}
            </Droppable>
          </Grid>
          <Grid
            item
            container
            direction="row"
            alignItems="flex-end"
            justify="flex-end"
          >
            <Grid item>
              {id === "todo" && (
                <Tooltip title="Add New Task">
                  <Fab
                    onClick={handleNewTaskButtonClick}
                    className={classes.addButton}
                    color="primary"
                    size="medium"
                  >
                    <img
                      src={AddIcon}
                      alt="add icon"
                      className={classes.addIcon}
                    />
                  </Fab>
                </Tooltip>
              )}
            </Grid>
          </Grid>
        </Grid>
      </Grid>
    </>
  );
};

Column.propTypes = {
  classes: PropTypes.shape({
    columnPaper: PropTypes.string,
    columnTitle: PropTypes.string,
    titleDivider: PropTypes.string,
    tasksListContainer: PropTypes.string,
    addButton: PropTypes.string,
    addIcon: PropTypes.string,
  }).isRequired,
  column: PropTypes.shape({
    title: PropTypes.string,
    id: PropTypes.string,
  }).isRequired,
  tasks: PropTypes.arrayOf(PropTypes.object).isRequired,
  openModal: PropTypes.func.isRequired,
};

const mapStateToProps = () => ({});

const mapDispatchToProps = (dispatch) => ({
  openModal: () => dispatch(actions.openModal()),
  closeModal: () => dispatch(actions.closeModal()),
});

export default compose(
  withStyles(styles),
  connect(mapStateToProps, mapDispatchToProps)
)(Column);
