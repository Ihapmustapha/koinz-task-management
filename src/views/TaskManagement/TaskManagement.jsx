import React, { useState, useEffect } from "react";
// prop-types
import PropTypes from "prop-types";
// redux
import { compose } from "redux";
import { connect } from "react-redux";
// beautiful-react-dnd
import { DragDropContext } from "react-beautiful-dnd";
// components
import { withStyles, Grid } from "@material-ui/core";
import { Column } from "./components";
// styles
import styles from "./styles";
// actions
import * as actions from "../../store/actions";
import { tasksMapper } from "../../mappers";

const TaskManagement = ({ fetchTasks, tasksList }) => {
  const [state, setState] = useState({
    tasks: {},
    columns: {},
    columnOrder: ["todo", "inProgress", "done"],
  });

  useEffect(() => {
    fetchTasks();
  }, [fetchTasks]);

  useEffect(() => {
    const mappedTasks = tasksMapper(tasksList);

    const columns = {
      todo: {
        id: "todo",
        title: "To Do",
        taskIds: [],
      },
      inProgress: {
        id: "inProgress",
        title: "In progress",
        taskIds: [],
      },
      done: {
        id: "done",
        title: "Done",
        taskIds: [],
      },
    };

    tasksList.forEach((task) => {
      if (task.id && task.taskStatus) {
        columns[task.taskStatus].taskIds.push(task.id);
      }
    });

    setState({ ...state, tasks: mappedTasks, columns });
  }, [tasksList]);

  const onDragEnd = (result) => {
    const { source, destination, draggableId } = result;
    const { columns } = state;

    // user dropped it outside all droppables
    if (!destination) return;
    // user didn't change item position, dragged and
    // -> dropped at the same possition
    const startColumn = { ...columns[source.droppableId] };
    const endColumn = { ...columns[destination.droppableId] };

    if (
      destination.droppableId === source.droppableId &&
      destination.index === source.index
    )
      return;

    const newStartTaskIds = [...startColumn.taskIds];
    const newEndColumnTaskIds = [...endColumn.taskIds];

    // reordering inside the same column
    newStartTaskIds.splice(source.index, 1);

    if (startColumn.id === endColumn.id) {
      newStartTaskIds.splice(destination.index, 0, draggableId);
      const newStartColumn = {
        ...startColumn,
        taskIds: newStartTaskIds,
      };
      setState({
        ...state,
        columns: { ...columns, [newStartColumn.id]: newStartColumn },
      });
      return;
    }

    // source !== destination
    // checking on accepted flows
    // todo => inProgress => done,
    // inProgress => todo || inProgress => done,
    // done => nowhere
    if (
      source.droppableId === "todo" &&
      destination.droppableId !== "inProgress"
    )
      return;
    if (source.droppableId === "done") return;
    // handling accepted flows
    const newStartColumn = {
      ...startColumn,
      taskIds: newStartTaskIds,
    };

    newEndColumnTaskIds.splice(destination.index, 0, draggableId);

    const newEndColumn = {
      ...endColumn,
      taskIds: newEndColumnTaskIds,
    };

    const newTaskStatus = newEndColumn.id;
    const taskId = draggableId;
    const newTasks = { ...state.tasks };
    const newTaskData = { ...newTasks[taskId] };
    newTaskData.taskStatus = newTaskStatus;
    newTasks[taskId] = newTaskData;

    setState({
      ...state,
      tasks: newTasks,
      columns: {
        ...columns,
        [newStartColumn.id]: newStartColumn,
        [newEndColumn.id]: newEndColumn,
      },
    });
  };

  return (
    <Grid
      container
      item
      xs={12}
      xl={12}
      direction="row"
      justify="space-evenly"
      alignItems="center"
    >
      {Object.entries(state.columns).length > 0 && (
        <DragDropContext onDragEnd={onDragEnd}>
          {state.columnOrder.map((columnId) => {
            const columnObj = state.columns[columnId];
            const arrOfTasks = columnObj.taskIds.map(
              (taskId) => state.tasks[taskId]
            );
            return (
              <Column key={columnId} column={columnObj} tasks={arrOfTasks} />
            );
          })}
        </DragDropContext>
      )}
    </Grid>
  );
};

TaskManagement.propTypes = {
  classes: PropTypes.shape({}).isRequired,
  fetchTasks: PropTypes.func.isRequired,
  tasksList: PropTypes.arrayOf(PropTypes.object).isRequired,
};

const mapStateToProps = (state) => {
  console.log(state.taskManagement);
  return {
    tasksList: state?.taskManagement?.tasksList || [],
  };
};
const mapDispatchToProps = (dispatch) => ({
  fetchTasks: () => dispatch(actions.fetchTasks()),
});

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  withStyles(styles)
)(TaskManagement);
