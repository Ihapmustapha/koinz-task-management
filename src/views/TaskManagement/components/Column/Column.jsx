import React from "react";
// prop-types
import PropTypes from "prop-types";
// redux
import { compose } from "redux";
import { connect } from "react-redux";
// react-beautiful-dnd
import { Droppable } from "react-beautiful-dnd";
// components
import {
  Grid,
  Paper,
  Divider,
  Tooltip,
  Fab,
  Typography,
  withStyles,
} from "@material-ui/core";
import Task from "../Task/Task";
import TaskForm from "../TaskForm/TaskForm";
import { ContainerModal } from "../../../../components";
// redux actions
import * as actions from "../../../../store/actions";
// styles
import styles from "../../styles";
// icons
import AddIcon from "../../../../assets/plus.svg";

const Column = ({
  column,
  tasks,
  classes,
  openModal,
  formType,
  updateTaskFormState,
  clearTaskFormState,
}) => {
  const { title, id } = column;

  const handleTaskCardClick = (selectedTaskId) => () => {
    const selectedTaskDetails = tasks.filter(
      (element) => element.id === selectedTaskId
    )[0];
    if (selectedTaskDetails) {
      clearTaskFormState();
      updateTaskFormState({ formType: "update", selectedTaskDetails });
      openModal();
    }
  };

  const handleNewTaskButtonClick = () => {
    clearTaskFormState();
    openModal();
  };

  return (
    <>
      {/* <Modal /> */}
      <ContainerModal
        title={formType === "add" ? "Add New Task" : "Update Task"}
      >
        <TaskForm />
      </ContainerModal>

      <Grid
        component={Paper}
        container
        item
        direction="column"
        justify="center"
        alignItems="center"
        elevation={3}
        className={classes.columnPaper}
      >
        <Grid
          item
          xs={12}
          lg={12}
          container
          direction="row"
          justify="center"
          alignItems="flex-start"
        >
          <Grid item xs={12}>
            <Typography
              variant="body2"
              data-testid="column-title-test-id"
              className={classes.columnTitle}
            >
              {title}
            </Typography>
            <Divider className={classes.titleDivider} />
          </Grid>
          <Grid item>
            <Droppable droppableId={id}>
              {(provided) => (
                <div
                  className={classes.tasksListContainer}
                  ref={provided.innerRef}
                  {...provided.droppableProps}
                >
                  {tasks.length > 0 &&
                    tasks.map(
                      (task, index) =>
                        task && (
                          <Task
                            key={task.id}
                            id={task.id}
                            index={index}
                            task={task}
                            onClick={handleTaskCardClick}
                          />
                        )
                    )}
                  {provided.placeholder}
                </div>
              )}
            </Droppable>
          </Grid>
          <Grid
            item
            container
            direction="row"
            alignItems="flex-end"
            justify="flex-end"
          >
            <Grid item>
              {id === "todo" && (
                <Tooltip title="Add New Task">
                  <Fab
                    onClick={handleNewTaskButtonClick}
                    className={classes.addButton}
                    color="primary"
                    size="medium"
                    data-testid="add-task-fab-test-id"
                  >
                    <img
                      src={AddIcon}
                      alt="add icon"
                      className={classes.addIcon}
                    />
                  </Fab>
                </Tooltip>
              )}
            </Grid>
          </Grid>
        </Grid>
      </Grid>
    </>
  );
};

Column.propTypes = {
  classes: PropTypes.shape({
    columnPaper: PropTypes.string,
    columnTitle: PropTypes.string,
    titleDivider: PropTypes.string,
    tasksListContainer: PropTypes.string,
    addButton: PropTypes.string,
    addIcon: PropTypes.string,
  }).isRequired,
  column: PropTypes.shape({
    title: PropTypes.string,
    id: PropTypes.string,
  }).isRequired,
  tasks: PropTypes.arrayOf(PropTypes.object).isRequired,
  openModal: PropTypes.func.isRequired,
  formType: PropTypes.string.isRequired,
  updateTaskFormState: PropTypes.func.isRequired,
  clearTaskFormState: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => {
  return {
    formType: state.taskForm.formType,
    selectedTaskDetails: state.taskForm.selectedTaskDetails,
  };
};

const mapDispatchToProps = (dispatch) => ({
  openModal: () => dispatch(actions.openModal()),
  closeModal: () => dispatch(actions.closeModal()),
  updateTaskFormState: (newState) =>
    dispatch(actions.updateTaskFormState(newState)),
  clearTaskFormState: () => dispatch(actions.clearTaskFormState()),
  handleTaskDescriptionChange: (value) =>
    dispatch(actions.handleTaskDescriptionChange(value)),
});

export default compose(
  withStyles(styles),
  connect(mapStateToProps, mapDispatchToProps)
)(Column);
