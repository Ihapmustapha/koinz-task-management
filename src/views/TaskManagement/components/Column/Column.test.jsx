import React from "react";
import { mount } from "enzyme";

import Column from "./Column";
import TaskManagement from "../../TaskManagement";

import { MockedProvider } from "../../../../test-utils";

describe("<Column /> with no props", () => {
  const getContainer = (props) =>
    mount(
      <MockedProvider initialState={{ modalState: { isOpen: false } }}>
        <TaskManagement>
          <Column {...props} />
        </TaskManagement>
      </MockedProvider>
    );

  const container = getContainer({});

  it("should match the snapshot", () => {
    expect(container.html()).toMatchSnapshot();
  });

  it("should contain typography title", () => {
    expect(container.find("#column-title-test-id").at(1).text()).toBe("To Do");
  });

  it("should contain Fab button to add new task when column is To Do", () => {
    expect(container.find("#add-task-fab-test-id").exists()).toBe(true);
  });
});
