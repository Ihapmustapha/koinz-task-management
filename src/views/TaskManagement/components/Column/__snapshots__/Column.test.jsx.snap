// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Column /> with no props should match the snapshot 1`] = `ReactWrapper {}`;
